File: client/DependencyGraph/DependencyGraph.cs
// Skeleton implementation written by Joe Zachary for CS 3500, September 2013.
// Version 1.1 (Fixed error in comment for RemoveDependency.)
// Version 1.2 - Daniel Kopta 
//               (Clarified meaning of dependent and dependee.)
//               (Clarified names in solution/project structure.)
//Version 1.3 - Tyler Wood
//               (Filled out skeleton code and added helper methods)

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SpreadsheetUtilities
{

    /// <summary>
    /// (s1,t1) is an ordered pair of strings
    /// t1 depends on s1; s1 must be evaluated before t1
    /// 
    /// A DependencyGraph can be modeled as a set of ordered pairs of strings.  Two ordered pairs
    /// (s1,t1) and (s2,t2) are considered equal if and only if s1 equals s2 and t1 equals t2.
    /// Recall that sets never contain duplicates.  If an attempt is made to add an element to a 
    /// set, and the element is already in the set, the set remains unchanged.
    /// 
    /// Given a DependencyGraph DG:
    /// 
    ///    (1) If s is a string, the set of all strings t such that (s,t) is in DG is called dependents(s).
    ///        (The set of things that depend on s)    
    ///        
    ///    (2) If s is a string, the set of all strings t such that (t,s) is in DG is called dependees(s).
    ///        (The set of things that s depends on) 
    //
    // For example, suppose DG = {("a", "b"), ("a", "c"), ("b", "d"), ("d", "d")}
    //     dependents("a") = {"b", "c"}
    //     dependents("b") = {"d"}
    //     dependents("c") = {}
    //     dependents("d") = {"d"}
    //     dependees("a") = {}
    //     dependees("b") = {"a"}
    //     dependees("c") = {"a"}
    //     dependees("d") = {"b", "d"}
    /// </summary>
    public class DependencyGraph
    {

        //the graph for the dependent relationships
        private Dictionary<String, HashSet<String>> dependentGraph;
        //the graph for the dependee relationships
        private Dictionary<String, HashSet<String>> dependeeGraph;
        //the number of dependencies
        private int size;

        /// <summary>
        /// Creates an empty DependencyGraph.
        /// </summary>
        public DependencyGraph()
        {
            dependentGraph = new Dictionary<string, HashSet<string>>();
            dependeeGraph = new Dictionary<string, HashSet<string>>();
            size = 0;
        }


        /// <summary>
        /// The number of ordered pairs in the DependencyGraph.
        /// </summary>
        public int Size
        {
            get { return size; }
        }


        /// <summary>
        /// The size of dependees(s).
        /// This property is an example of an indexer.  If dg is a DependencyGraph, you would
        /// invoke it like this:
        /// dg["a"]
        /// It should return the size of dependees("a")
        /// </summary>
        public int this[string s]
        {
            get 
            {
                //As long as dependentgraph contains the key, return the size of the Hashset from that key. Otherwise return 0
                if (dependeeGraph.TryGetValue(s, out HashSet<string> result))
                {
                    return result.Count;
                }
                return 0; 
            }
        }


        /// <summary>
        /// Reports whether dependents(s) is non-empty.
        /// </summary>
        public bool HasDependents(string s)
        {
            if (!dependentGraph.ContainsKey(s) || dependentGraph[s].Count == 0)
            {
                return false;
            }
            return true;
            
        }


        /// <summary>
        /// Reports whether dependees(s) is non-empty.
        /// </summary>
        public bool HasDependees(string s)
        {
            if (this[s] == 0)
            {
                return false;
            }
            return true;

        }


        /// <summary>
        /// Enumerates dependents(s).
        /// </summary>
        public IEnumerable<string> GetDependents(string s)
        {
            //As long as the dependentgraph contains the string s, return the hashset from that dependentgraph. Otherwise return an empty list
            if (!dependentGraph.ContainsKey(s))
            {
                return new HashSet<string>();
            }
            return dependentGraph[s];
            
        }

        /// <summary>
        /// Enumerates dependees(s).
        /// </summary>
        public IEnumerable<string> GetDependees(string s)
        {
            //As long as the dependeegraph contains the string s, return the hashset from that dependeegraph. Otherwise return an empty list
            if (!dependeeGraph.ContainsKey(s))
            {
                return new HashSet<string>();
            }
            return dependeeGraph[s];
        }


        /// <summary>
        /// <para>Adds the ordered pair (s,t), if it doesn't exist</para>
        /// 
        /// <para>This should be thought of as:</para>   
        /// 
        ///   t depends on s
        ///
        /// </summary>
        /// <param name="s"> s must be evaluated first. T depends on S</param>
        /// <param name="t"> t cannot be evaluated until s is</param>        /// 
        public void AddDependency(string s, string t)
        {
            AddDependentHelper(s, t);
            AddDependeeHelper(s, t);
            
            
        }

        /// <summary>
        /// Assists with the adding of s and t to the dependent graph
        /// </summary>
        /// <param name="s"> the dependent of the ordered pair</param>
        /// <param name="t"> the dependee of the ordered pair</param>
        private void AddDependentHelper(string s, string t)
        {
            //Check to see if the dependent graph already contains s, then t. If it does, do nothing, if only s, then add t and increase size
            if (dependentGraph.ContainsKey(s))
            {
                if (!dependentGraph[s].Contains(t))
                {
                    dependentGraph[s].Add(t);
                    size++;
                }
            }

            //If s is not in the dependentgraph, then add both s and t
            else
            {
                dependentGraph.Add(s, new HashSet<string>());
                dependentGraph[s].Add(t);
                size++;
            }

            //Check to see if t is in the dependentgraph, if not then add it
            if (!dependentGraph.ContainsKey(t))
            {
                dependentGraph.Add(t, new HashSet<String>());
            }
        }

        /// <summary>
        /// Assists with the adding of s and t to the dependee graph
        /// </summary>
        /// <param name="s"> the dependent of the ordered pair</param>
        /// <param name="t"> the dependee of the ordered pair</param>
        private void AddDependeeHelper(string s, string t)
        {
            //Check to see if the dependee graph already contains t, then s. If it does, do nothing, if only t, then add s
            if (dependeeGraph.ContainsKey(t))
            {
                if (!dependeeGraph[t].Contains(s))
                {
                    dependeeGraph[t].Add(s);
                }

            }

            //Otherwise, add t and s to the dependee graph
            else
            {
                dependeeGraph.Add(t, new HashSet<string>());
                dependeeGraph[t].Add(s);
            }

            //Check to see if s is in the dependee graph. If not, add it
            if (!dependeeGraph.ContainsKey(s))
            {
                dependeeGraph.Add(s, new HashSet<String>());
            }
        }


        /// <summary>
        /// Removes the ordered pair (s,t), if it exists
        /// </summary>
        /// <param name="s"> the dependent of the ordered pair</param>
        /// <param name="t"> the dependee of the ordered pair</param>
        public void RemoveDependency(string s, string t)
        {
            //Check to see if s, then t is in the dependent graph. If so, then remove it
            if (dependentGraph.ContainsKey(s))
            {
                if (dependentGraph[s].Contains(t))
                {
                    helpRemoveDependent(s, t);
                }
            }

            //Check to see if t, then s is in the dependee graph. If so, then remove it
            if (dependeeGraph.ContainsKey(t))
            {
                if (dependeeGraph[t].Contains(s))
                {
                    helpRemoveDependee(s, t);
                }
            }
        }

        /// <summary>
        /// Helper method that removes the dependent relationship from the graph, and the two points if they no longer have any dependents
        /// </summary>
        /// <param name="s"> the dependent of the ordered pair</param>
        /// <param name="t"> the dependee of the ordered pair</param>
        private void helpRemoveDependent(string s, string t)
        {
            //Remove the (s, t) pair from the dependent graph, and decrement size
            dependentGraph[s].Remove(t);
            size--;

            //If t's dependent set is empty, remove it from the graph
            if (dependentGraph.ContainsKey(t) && dependentGraph[t].Count == 0)
            {
                dependentGraph.Remove(t);
            }

            //If s' dependent set is now empty, remove it from the graph
            if (dependentGraph[s].Count == 0)
            {
                dependentGraph.Remove(s);
            }
        }

        /// <summary>
        /// Helper method that removes the dependee relationship from the graph, and the two points if they no longer have any dependees
        /// </summary>
        /// <param name="s"> the dependent of the ordered pair</param>
        /// <param name="t"> the dependee of the ordered pair</param>
        private void helpRemoveDependee(string s, string t)
        {
            //Remove the pair (t, s) from the dependee graph
            dependeeGraph[t].Remove(s);

            //If s' dependee set is empty, remove it from the graph
            if(dependeeGraph.ContainsKey(s) && dependeeGraph[s].Count == 0)
            {
                dependeeGraph.Remove(s);
            }

            //If t's dependee set is now empty, remove it from the graph
            if(dependeeGraph[t].Count == 0)
            {
                dependeeGraph.Remove(t);
            }

        }


        /// <summary>
        /// Removes all existing ordered pairs of the form (s,r).  Then, for each
        /// t in newDependents, adds the ordered pair (s,t).
        /// </summary>
        public void ReplaceDependents(string s, IEnumerable<string> newDependents)
        {
            //Create a new HashSet to ensure that no enumerating is happening while the original set is being modified
            HashSet<string> remove = new HashSet<string>(GetDependents(s));

            //Go through each string within the new hashset, and remove the ordered pair (s, t) from the graph
            foreach (string t in remove)
            {
                RemoveDependency(s, t);
            }

            //Add each new dependency from the newDependents IEnumerable
            foreach (string t in newDependents)
            {
                AddDependency(s, t);
            }
        }


        /// <summary>
        /// Removes all existing ordered pairs of the form (r,s).  Then, for each 
        /// t in newDependees, adds the ordered pair (t,s).
        /// </summary>
        public void ReplaceDependees(string s, IEnumerable<string> newDependees)
        {
            //Create a new HashSet to ensure that no enumerating is happening while the original set is being modified
            HashSet<string> remove = new HashSet<string>(GetDependees(s));

            //Go through each string from the new hashset, and remove the ordered pair (t, s) from the graph
            foreach(string t in remove)
            {
                RemoveDependency(t, s);
            }

            //Add each new dependency from the newDependents IEnumerable
            foreach(string t in newDependees)
            {
                AddDependency(t, s);
            }
        }

    }

}

File: server/DependencyGraph.cpp
/* COCOPRO Networked Spreadsheet Final Project 
 * 
 * spreadsheet DependencyGraph class definitions
 */

#include "DependencyGraph.h"
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <iostream> 

DependencyGraph::DependencyGraph ()
{
  dependentGraph = {};
  dependeeGraph = {};
  size = 0;
}

int DependencyGraph::getSize () const 
{
  return size;
}

bool DependencyGraph::hasDependents(std::string s)
{
  if (!dependentContains(s) || dependentGraph[s].size() == 0)
    return false;
  return true;
}

bool DependencyGraph::hasDependees(std::string s)
{
  if (!dependeeContains(s) || dependeeGraph[s].size() == 0)
    return false;
  return true;
}

std::vector<std::string> DependencyGraph::getDependents(std::string s)
{
  //As long as the dependentgraph contains the string s, return the hashset from that dependentgraph. Otherwise return an empty list
  if (!dependentContains(s))
    return std::vector<std::string>();
  std::vector<std::string> output(dependentGraph[s].begin(), dependentGraph[s].end());
  return output;
}

std::vector<std::string> DependencyGraph::getDependees(std::string s)
{
  //As long as the dependeegraph contains the string s, return the hashset from that dependeegraph. Otherwise return an empty list
  if (!dependeeContains(s))
    return std::vector<std::string>();
  std::vector<std::string> output(dependeeGraph[s].begin(), dependeeGraph[s].end());
  return output;
}

void DependencyGraph::addDependency(std::string s, std::string t)
{
  addDependentHelper(s, t);
  addDependeeHelper(s, t);
}

void DependencyGraph::removeDependency(std::string s, std::string t)
{
  //Check to see if s, then t is in the dependent graph. If so, then remove it
  if (dependentContains(s))
  {
    if (dependentGraph[s].find(t) != dependentGraph[s].end())
    {
      helpRemoveDependent(s, t);
    }
  }

  //Check to see if t, then s is in the dependee graph. If so, then remove it
  if (dependeeContains(t))
  {
    if (dependeeGraph[t].find(t) != dependeeGraph[s].end())
    {
      helpRemoveDependee(s, t);
    }
  }
}

void DependencyGraph::replaceDependents(std::string s, std::vector<std::string> dependents)
{
  //Create a  HashSet to ensure that no enumerating is happening while the original set is being modified
  std::vector<std::string> remove(getDependents(s));

  //Go through each string within the  hashset, and remove the ordered pair (s, t) from the graph
  
  for (int i = 0; i < remove.size(); i++)
  {
    removeDependency(s, remove[i]);
  }

  //Add each  dependency from the Dependents vector
  for (int i = 0; i < dependents.size(); i++)
  {
	addDependency(s, dependents[i]);
  }
  
  
}

void DependencyGraph::replaceDependees(std::string s, std::vector<std::string> dependees)
{
  //Create a  HashSet to ensure that no enumerating is happening while the original set is being modified
  std::vector<std::string> remove(getDependees(s));
  //Go through each string from the  hashset, and remove the ordered pair (t, s) from the graph
  for (int i = 0; i < remove.size(); i++)
  {
    removeDependency(remove[i], s);
  }

if(dependees.size() != 0)
{
  //Add each  dependency from the Dependents vector
  for (int i = 0; i < dependees.size(); i++)
  {
    addDependency(dependees[i], s);
  }
}
  
}

void DependencyGraph::addDependentHelper(std::string s, std::string t)
{
  //Check to see if the dependent graph already contains s, then t. If it does, do nothing, if only s, then add t and increase size
  if (dependentContains(s))
  {
    if (dependentGraph[s].find(t) == dependentGraph[s].end())
    {
      dependentGraph[s].insert(t);
      size++;
    }
  }

  //If s is not in the dependentgraph, then add both s and t
  else
  {
    dependentGraph.insert(std::pair<std::string, std::unordered_set<std::string>>(s, std::unordered_set<std::string>()));
    dependentGraph[s].insert(t);
    size++;
  }

  //Check to see if t is in the dependentgraph, if not then add it
  if (!dependentContains(t))
  {
    dependentGraph.insert(std::pair<std::string, std::unordered_set<std::string>>(t, std::unordered_set<std::string>()));
  }
}

void DependencyGraph::addDependeeHelper(std::string s, std::string t)
{
  //Check to see if the dependee graph already contains t, then s. If it does, do nothing, if only t, then add s
  if (dependeeContains(t))
  {
    if (dependeeGraph[t].find(t) == dependeeGraph[s].end())
    {
      dependeeGraph[t].insert(s);
    }

  }

  //Otherwise, add t and s to the dependee graph
  else
  {
    dependeeGraph.insert(std::pair<std::string, std::unordered_set<std::string>>(t,  std::unordered_set<std::string>()));
    dependeeGraph[t].insert(s);
  }

  //Check to see if s is in the dependee graph. If not, add it
  if (!dependeeContains(s))
  {
    dependeeGraph.insert(std::pair<std::string, std::unordered_set<std::string>>(s, std::unordered_set<std::string>()));
  }
}

void DependencyGraph::helpRemoveDependent(std::string s, std::string t)
{
  //Remove the pair (t, s) from the dependent graph
  dependentGraph[t].erase(s);

  //If s' dependent set is empty, remove it from the graph
  if(dependentContains(s) && dependentGraph[s].size() == 0)
  {
    dependentGraph.erase(s);
  }

  //If t's dependent set is now empty, remove it from the graph
  if(dependentGraph[t].size() == 0)
  {
    dependentGraph.erase(t);
  }
}

void DependencyGraph::helpRemoveDependee(std::string s, std::string t)
{
  //Remove the pair (t, s) from the dependee graph
  dependeeGraph[t].erase(s);

  //If s' dependee set is empty, remove it from the graph
  if(dependeeContains(s) && dependeeGraph[s].size() == 0)
  {
    dependeeGraph.erase(s);
  }

  //If t's dependee set is now empty, remove it from the graph
  if(dependeeGraph[t].size() == 0)
  {
    dependeeGraph.erase(t);
  }
}

bool DependencyGraph::dependentContains (std::string s)
{
  auto search = dependentGraph.find(s);
  return (search != dependentGraph.end());
}

bool DependencyGraph::dependeeContains (std::string s)
{
  auto search = dependeeGraph.find(s);
  return (search != dependeeGraph.end());
}

